<a href="https://colab.research.google.com/github/rajat2019/DataScienceBasics/blob/main/AssociationRuleLearning/Apriori.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Apriori

**Apriori intution:**

* Step 1: Set a minimum support and confidence
* Step 2: Take all the subsetes in transaction having higher support than miimum support
* Step 3: Take ll the rules of theses subste having higher confidence that minimum confidence
* Step 4: Sort the rules by decreasing lift lift (M1-->M2) = confidence (M1-->M2)/ support(M2)





Apriori is a basic approach but in production it much more complex

* People who bought this also bought these!
* Used in ecommerce and reltail 
* Here we are looking for best association rules 

## Importing the libraries

In [ ]:
!pip install apyori  #apriori isn't available in scikit learn 

  Created wheel for apyori: filename=apyori-1.1.2-py3-none-any.whl size=5974 sha256=3d2383bb141fe5a7bf76a73f831ff13848c11e424ec277c4567c99c38c469ad4
  Stored in directory: /root/.cache/pip/wheels/cb/f6/e1/57973c631d27efd1a2f375bd6a83b2a616c4021f24aab84080
Successfully built apyori


In [ ]:
import numpy as np
import matplotlib.pyplot as plt
import pandas as pd

## Data Preprocessing

In [ ]:
dataset = pd.read_csv('Market_Basket_Optimisation.csv', header = None) # header = None ...since there is no header as earlier datasets

transactions = [] #dataset as a list of transactions for apriori function/method in apyori library

for i in range(0, 7501): #uppper limit is excluded
  transactions.append([str(dataset.values[i,j]) for j in range(0, 20)]) # we are creating a list of lists....good examples of one line for loop
  #for apriori model all the items must be string...that's why we used str

## Training the Apriori model on the dataset

In [ ]:
from apyori import apriori  # from apyori package that we installed importe the apriori function 
rules = apriori(transactions = transactions, min_support = 0.003, min_confidence = 0.2, min_lift = 3, min_length = 2, max_length = 2)

First argument is data...which is transactions. It should be a list of strings. 

min_support: minimum support is 7 * 3 / 7501 (7501 which is total transaction in the week). This goes well with the current scenario

support considered by the model is in fact, sets of items/total items


min_confidence: minimum condidence ...rule of thumb ...0.8 ....but we will use 0.2
0.8 was way too high...so we have to reduce the confidence to 0.2 since using 
0.8 you will not get any rule

min_lift: minimum lift ....atleast 3 times per day ..so 21 times in a week...lift below 3 makes the rule not relevant....

min_length and max_length are based on business problem

min length ...rules must have atleast 2 products 
max length ....max number in a rule..we want to to simpliy this problem...so lets keep it at 2

...IF buy 2 products and get one free...then min and max length would be 3



## Visualising the results

### Displaying the first results coming directly from the output of the apriori function

In [ ]:
results = list(rules)

In [ ]:
results

[RelationRecord(items=frozenset({'light cream', 'chicken'}), support=0.004532728969470737, ordered_statistics=[OrderedStatistic(items_base=frozenset({'light cream'}), items_add=frozenset({'chicken'}), confidence=0.29059829059829057, lift=4.84395061728395)]),
 RelationRecord(items=frozenset({'mushroom cream sauce', 'escalope'}), support=0.005732568990801226, ordered_statistics=[OrderedStatistic(items_base=frozenset({'mushroom cream sauce'}), items_add=frozenset({'escalope'}), confidence=0.3006993006993007, lift=3.790832696715049)]),
 RelationRecord(items=frozenset({'pasta', 'escalope'}), support=0.005865884548726837, ordered_statistics=[OrderedStatistic(items_base=frozenset({'pasta'}), items_add=frozenset({'escalope'}), confidence=0.3728813559322034, lift=4.700811850163794)]),
 RelationRecord(items=frozenset({'fromage blanc', 'honey'}), support=0.003332888948140248, ordered_statistics=[OrderedStatistic(items_base=frozenset({'fromage blanc'}), items_add=frozenset({'honey'}), confidence=0

In [ ]:
results[0][2]

[OrderedStatistic(items_base=frozenset({'light cream'}), items_add=frozenset({'chicken'}), confidence=0.29059829059829057, lift=4.84395061728395)]

### Putting the results well organised into a Pandas DataFrame

In [ ]:
def inspect(results):
    lhs         = [tuple(result[2][0][0])[0] for result in results] 
    # index 0: "RelationRecord(items=frozenset({'light cream', 'chicken'})" 
    # index 1: support=0.004532728969470737
    # index 2: ordered_statistics=[OrderedStatistic(items_base=frozenset({'light cream'}), items_add=frozenset({'chicken'}), confidence=0.29059829059829057, lift=4.84395061728395)]
    ## from element of index 2 we take index 0 which is [OrderedStatistic(items_base=frozenset({'light cream'}), items_add=frozenset({'chicken'}), confidence=0.29059829059829057, lift=4.84395061728395)]
    ### from this index 2/0 we take light cream using index 0
    rhs         = [tuple(result[2][0][1])[0] for result in results]
    supports    = [result[1] for result in results]
    confidences = [result[2][0][2] for result in results]
    lifts       = [result[2][0][3] for result in results]
    return list(zip(lhs, rhs, supports, confidences, lifts))
resultsinDataFrame = pd.DataFrame(inspect(results), columns = ['Left Hand Side', 'Right Hand Side', 'Support', 'Confidence', 'Lift'])
#this code from stack overflow




```
>>> numbers = [1, 2, 3]
>>> letters = ['a', 'b', 'c']
>>> zipped = zip(numbers, letters)
>>> zipped  # Holds an iterator object
<zip object at 0x7fa4831153c8>
>>> type(zipped)
<class 'zip'>
>>> list(zipped)
[(1, 'a'), (2, 'b'), (3, 'c')]
```


Here, you use zip(numbers, letters) to create an iterator that produces tuples of the form (x, y). In this case, the x values are taken from numbers and the y values are taken from letters. Notice how the Python zip() function returns an iterator. To retrieve the final list object, you need to use list() to consume the iterator.

If you’re working with sequences like lists, tuples, or strings, then your iterables are guaranteed to be evaluated from left to right. This means that the resulting list of tuples will take the form [(numbers[0], letters[0]), (numbers[1], letters[1]),..., (numbers[n], letters[n])]

### Displaying the results non sorted

In [ ]:
resultsinDataFrame

Left Hand Side Right Hand Side   Support  Confidence      Lift
0           light cream         chicken  0.004533    0.290598  4.843951
1  mushroom cream sauce        escalope  0.005733    0.300699  3.790833
2                 pasta        escalope  0.005866    0.372881  4.700812
3         fromage blanc           honey  0.003333    0.245098  5.164271
4         herb & pepper     ground beef  0.015998    0.323450  3.291994
5          tomato sauce     ground beef  0.005333    0.377358  3.840659
6           light cream       olive oil  0.003200    0.205128  3.114710
7     whole wheat pasta       olive oil  0.007999    0.271493  4.122410
8                 pasta          shrimp  0.005066    0.322034  4.506672

### Displaying the results sorted by descending lifts

In [ ]:
resultsinDataFrame.nlargest(n = 10, columns = 'Lift')

Left Hand Side Right Hand Side   Support  Confidence      Lift
3         fromage blanc           honey  0.003333    0.245098  5.164271
0           light cream         chicken  0.004533    0.290598  4.843951
2                 pasta        escalope  0.005866    0.372881  4.700812
8                 pasta          shrimp  0.005066    0.322034  4.506672
7     whole wheat pasta       olive oil  0.007999    0.271493  4.122410
5          tomato sauce     ground beef  0.005333    0.377358  3.840659
1  mushroom cream sauce        escalope  0.005733    0.300699  3.790833
4         herb & pepper     ground beef  0.015998    0.323450  3.291994
6           light cream       olive oil  0.003200    0.205128  3.114710

The "support" shown in the output is calculated by (# transaction containing item 1 & item 2)/(# of transactions). This is different than what was taught and explains why you can't take the output "support" and "confidence" to calculate the "lift".

A specific example using the light cream & chicken to illustrate what I'm saying. If you review the data set you'll see:



* Number of light cream transactions = 117

* Number of chicken transactions = 450

* Number of light cream & chicken transactions = 34

* Number of total transactions = 7501


Support (Chicken) = 450/7501 = 0.05999 *this is not shown in the output

Confidence (Chicken -> LC) = 34 /117 = 0.2905983

Lift = 0.05999/0.2905983 = 4.843951

Support (LC & Chicken) = 34/7501=0.004533 *what's shown in the output."

